src/main/sensors/compass.c

```diff
@@ -61,32 +60,44 @@ extern uint32_t currentTime; // FIXME dependency on global variable, pass it in
 int16_t magADCRaw[XYZ_AXIS_COUNT];
 int32_t magADC[XYZ_AXIS_COUNT];
 sensor_align_e magAlign = 0;
+
 #ifdef MAG
+
 static uint8_t magInit = 0;
 
-void compassInit(void)
+bool compassInit(void)
 {
     // initialize and calibration. turn on led during mag calibration (calibration routine blinks it)
     LED1_ON;
-    mag.init();
+    const bool ret = mag.init();
     LED1_OFF;
-    magInit = 1;
+    if (ret) {
+        magInit = 1;
+    }
+    return ret;
 }
 
 void updateCompass(flightDynamicsTrims_t *magZero)
 {
     static uint32_t tCal = 0;
     static flightDynamicsTrims_t magZeroTempMin;
     static flightDynamicsTrims_t magZeroTempMax;
-    uint32_t axis;
 
     mag.read(magADCRaw);
-    for (axis = 0; axis < XYZ_AXIS_COUNT; axis++) magADC[axis] = magADCRaw[axis];  // int32_t copy to work with
+    if (!mag.read(magADCRaw)) {
```

Comment:
Sorry, I don't see double read in old code ...

As for return value - IMO it's OK for `mag.read` to return undefined value (avoid zeroing return value inside / return what's available). Caller shall handle that. It will slightly help with debugging and will remove code duplication ...